Add Docker to Foundry

abdulrabbani00 opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

This PR is co-authored with @dmfxyz - He provided the final `Dockerfile`.

I created this PR for the following reasons:
1. Utilize a foundry image in my own development.
2. Utilize a devcontainer if I need to develop any features of foundry.
3. I would like the foundry team to maintain the image if possible. Otherwise, I can maintain it myself.

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->
A simple Dockerfile, GitHub Actions Pipeline, and a devcontainer folder.

# Caution

The CI/CD pipeline seems to be working fine but I can't push it to Dockerhub due to "authorization scope." I don't want to spent time debugging the credentials for deploying to Dockerhub.

# Testing
I recommend testing the CI/CD pipeline prior to merging it to `master`.

mattsse commented on .github/workflows/release-container.yml:
```diff
@@ -0,0 +1,32 @@
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Login to DockerHub
+        uses: docker/login-action@v1
```
how about publishing to ghcr instead?

I used this in the past

```yaml
name: docker

on:
  push:
    branches: master

jobs:
  docker:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v1
      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v1
      - name: Login to Github Registry
        uses: docker/login-action@v1
        with:
          registry: ghcr.io
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}
      - name: PrepareReg Names
        run: echo IMAGE_REPOSITORY=$(echo ${{ github.repository }} | tr '[:upper:]' '[:lower:]') >> $GITHUB_ENV
      - name: Build and push
        id: docker_build
        uses: docker/build-push-action@v2
        with:
          push: ${{  github.event_name == 'push' }}
          tags: ghcr.io/gakonst/foundry

```

mattsse commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+WORKDIR /root
+ADD . .
+
+ENV SHELL "/bin/bash"
+RUN ["/bin/bash", "foundryup/install"]
+
+ENV PATH "$PATH:/root/.foundry/bin/"
+RUN echo "export PATH=/new/path:${PATH}" >> $HOME/.bashrc; \
```
what if we install the latest head instead of using foundryup?

mattsse reviewed (commented):
thanks, two points

* we should probably deploy to ghcr 
* I think we should rather install the latest head in the dockerfile

mattsse commented:
The official rust image also has some examples https://hub.docker.com/_/rust

abdulrabbani00 commented:
@mattsse - Noticing the following error when compiling: 

```
error: could not compile `foundry-cli`

Caused by:
  process didn't exit successfully: `rustc --crate-name cast --edition=2021 cli/src/cast.rs --error-format=json --json=diagnostic-rendered-ansi,future-incompat --crate-type bin --emit=dep-info,link -C opt-level=z -C panic=abort -C lto -C codegen-units=1 -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="rustls"' --cfg 'feature="sputnik"' --cfg 'feature="sputnik-evm"' -C metadata=c24ea9286d38eee2 -C extra-filename=-c24ea9286d38eee2 --out-dir /tmp/cargo-install3Uynzk/release/deps -L dependency=/tmp/cargo-install3Uynzk/release/deps --extern ansi_term=/tmp/cargo-install3Uynzk/release/deps/libansi_term-2304781d38eb3624.rlib --extern atty=/tmp/cargo-install3Uynzk/release/deps/libatty-2b50803019812599.rlib --extern cast=/tmp/cargo-install3Uynzk/release/deps/libcast-22595ea7d1ca326c.rlib --extern clap=/tmp/cargo-install3Uynzk/release/deps/libclap-ff37a1114f5cd7e1.rlib --extern clap_complete=/tmp/cargo-install3Uynzk/release/deps/libclap_complete-d8c4e6a78fbd8568.rlib --extern color_eyre=/tmp/cargo-install3Uynzk/release/deps/libcolor_eyre-dbad56b48104b659.rlib --extern console=/tmp/cargo-install3Uynzk/release/deps/libconsole-9ee7a12c895c7718.rlib --extern dunce=/tmp/cargo-install3Uynzk/release/deps/libdunce-99bd2858659effcd.rlib --extern ethers=/tmp/cargo-install3Uynzk/release/deps/libethers-7e1f1eb64bcf5410.rlib --extern sputnik=/tmp/cargo-install3Uynzk/release/deps/libevm-05afc08d80c3e2a0.rlib --extern evm_adapters=/tmp/cargo-install3Uynzk/release/deps/libevm_adapters-aa805e7372009d09.rlib --extern eyre=/tmp/cargo-install3Uynzk/release/deps/libeyre-17387a4470c0bbeb.rlib --extern forge=/tmp/cargo-install3Uynzk/release/deps/libforge-9c853b8b099800e7.rlib --extern forge_fmt=/tmp/cargo-install3Uynzk/release/deps/libforge_fmt-51780dfa31fcb34c.rlib --extern foundry_config=/tmp/cargo-install3Uynzk/release/deps/libfoundry_config-37ec376bdf4f516f.rlib --extern foundry_utils=/tmp/cargo-install3Uynzk/release/deps/libfoundry_utils-99245395a0c68544.rlib --extern futures=/tmp/cargo-install3Uynzk/release/deps/libfutures-ec5be1edb92ebe53.rlib --extern glob=/tmp/cargo-install3Uynzk/release/deps/libglob-96e55c85069b1667.rlib --extern hex=/tmp/cargo-install3Uynzk/release/deps/libhex-4b6a955ed4498af7.rlib --extern locate_cargo_manifest=/tmp/cargo-install3Uynzk/release/deps/liblocate_cargo_manifest-04cf1e9c207dc0ee.rlib --extern once_cell=/tmp/cargo-install3Uynzk/release/deps/libonce_cell-3a6d324dcc1a86f4.rlib --extern proptest=/tmp/cargo-install3Uynzk/release/deps/libproptest-d5ec1e9d6c29636a.rlib --extern rayon=/tmp/cargo-install3Uynzk/release/deps/librayon-7c58ededa153df70.rlib --extern regex=/tmp/cargo-install3Uynzk/release/deps/libregex-d1ea232f82b4b232.rlib --extern rpassword=/tmp/cargo-install3Uynzk/release/deps/librpassword-2af9315386b9212b.rlib --extern rustc_hex=/tmp/cargo-install3Uynzk/release/deps/librustc_hex-7ff161deb5a27052.rlib --extern semver=/tmp/cargo-install3Uynzk/release/deps/libsemver-e70879db2fea9bbb.rlib --extern serde=/tmp/cargo-install3Uynzk/release/deps/libserde-a952f9b94115863d.rlib --extern serde_json=/tmp/cargo-install3Uynzk/release/deps/libserde_json-ea2fb15ca9e710a8.rlib --extern similar=/tmp/cargo-install3Uynzk/release/deps/libsimilar-484d36495a81f210.rlib --extern solang_parser=/tmp/cargo-install3Uynzk/release/deps/libsolang_parser-dc6232bf1bc54a0a.rlib --extern tokio=/tmp/cargo-install3Uynzk/release/deps/libtokio-6a7a83dd87b790bc.rlib --extern tracing=/tmp/cargo-install3Uynzk/release/deps/libtracing-f83934623b1dfd26.rlib --extern tracing_subscriber=/tmp/cargo-install3Uynzk/release/deps/libtracing_subscriber-d86a287943cea664.rlib --extern ui=/tmp/cargo-install3Uynzk/release/deps/libui-fce0324a5815a6ed.rlib --extern walkdir=/tmp/cargo-install3Uynzk/release/deps/libwalkdir-647610b184c56b15.rlib --extern watchexec=/tmp/cargo-install3Uynzk/release/deps/libwatchexec-3b2dd1a7b45a508e.rlib -L native=/tmp/cargo-install3Uynzk/release/build/ring-b4d8f313d6cbb62f/out -L native=/tmp/cargo-install3Uynzk/release/build/hidapi-rusb-5d9413fe41a6beb5/out -L native=/tmp/cargo-install3Uynzk/release/build/libusb1-sys-10bfa4266d309ab3/out -L native=/tmp/cargo-install3Uynzk/release/build/libgit2-sys-e9e0778ee5f30378/out/build` (signal: 9, SIGKILL: kill)
warning: build failed, waiting for other jobs to finish...
```

abdulrabbani00 commented:
@mattsse - I have updated the CI/CD pipeline to use the registry as advised. I also attempted to build the image on Debian and CentOS but keep running into build errors. My recommendation is to test what we have now and get it merged. 

Down the road, someone can update the build process to build from the source if they choose.

abdulrabbani00 commented:
@mattsse , @gakonst - Any thoughts fellas? Would love to integrate this maintained image in some of my work tomorrow. Cheers.

mattsse commented:
this error indicates that foundryup compiles from source rather than pulling latest nightly? don't really know how foundry/install works but @onbjerg can probably here.

installing fro source should be as simple as 

```yaml
FROM rust

WORKDIR /usr/src/foundry
COPY . .

RUN cd cli && cargo install --path .
```

abdulrabbani00 commented:
```
#7 474.4    Compiling forge v0.1.0 (/usr/src/foundry/forge)
#7 527.2 warning: use of deprecated associated function `clap::CommandFactory::into_app`: Replaced with `CommandFactory::command
#7 527.2   --> cli/src/forge.rs:77:40
#
[2022-03-13T20:48:08.001Z] 7 527.2    |
[2022-03-13T20:48:08.001Z] 
#7 527.2 77 |             generate(shell, &mut Opts::into_app(), "forge", &mut std::io::stdout())
#7 527.2    |                                        ^^^^^^^^
#7 527.2    |
#7 527.2    = note: `#[warn(deprecated)]` on by default
#7 527.2 
#7 531.3 warning: use of deprecated associated function `clap::CommandFactory::into_app`: Replaced with `CommandFactory::command
#7 531.3    --> cli/src/cast.rs:727:40
#7 531.3     |
#7 531.3 727 |             generate(shell, &mut Opts::into_app(), "cast", &mut std::io::stdout())
#7 531.3     |                                        ^^^^^^^^
#7 531.3     |
#7 531.3     = note: `#[warn(deprecated)]` on by default
#7 531.3 
[2022-03-13T20:48:08.001Z] #7 723.1 warning: `foundry-cli` (bin "cast") generated 1 warning
#7 723.1 error: could not compile `foundry-cli`; 1 warning emitted
#7 723.1 
#7 723.1 Caused by:
#7 723.2   process didn't exit successfully: `rustc --crate-name cast --edition=2021 cli/src/cast.rs --error-format=json --json=diagnostic-rendered-ansi,future-incompat --crate-type bin --emit=dep-info,link -C opt-level=z -C panic=abort -C lto -C codegen-units=1 -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="rustls"' --cfg 'feature="sputnik"' --cfg 'feature="sputnik-evm"' -C metadata=847afcd56729a66d -C extra-filename=-847afcd56729a66d --out-dir /usr/src/foundry/target/release/deps -L dependency=/usr/src/foundry/target/release/deps --extern ansi_term=/usr/src/foundry/target/release/deps/libansi_term-2304781d38eb3624.rlib --extern atty=/usr/src/foundry/target/release/deps/libatty-be809fc78be2c052.rlib --extern cast=/usr/src/foundry/target/release/deps/libcast-99c3379508d83ba6.rlib --extern clap=/usr/src/foundry/target/release/deps/libcla
[2022-03-13T20:48:08.002Z] p-1d7327f01db0c4cf.rlib --extern clap_complete=/usr/src/foundry/target/release/deps/libclap_complete-cb348ee64b72417e.rlib --extern color_eyre=/usr/src/foundry/target/release/deps/libcolor_eyre-02a59259b0a7cb90.rlib --extern console=/usr/src/foundry/target/release/deps/libconsole-a5c7b4ecb77f7473.rlib --extern dunce=/usr/src/foundry/target/release/deps/libdunce-99bd2858659effcd.rlib --extern ethers=/usr/src/foundry/target/release/deps/libethers-26ab86522cbfef97.rlib --extern sputnik=/usr/src/foundry/target/release/deps/libevm-60c4244b617c66a7.rlib --extern evm_adapters=/usr/src/foundry/target/release/deps/libevm_adapters-d9f6001185893e9b.rlib --extern eyre=/usr/src/foundry/target/release/deps/libeyre-5a3967f052e3d34e.rlib --extern forge=/usr/src/foundry/target/release/deps/libforge-f2e1a51f3fe0339c.rlib --extern forge_fmt=/usr/src/foundry/target/release/deps/libforge_fmt-3b888e3b1b12888a.rlib --extern foundry_config=/usr/src/foundry/target/release/deps/libfoundry_config-d20e9df3c3b773e1.rlib --extern foundry_
[2022-03-13T20:48:08.004Z] utils=/usr/src/foundry/target/release/deps/libfoundry_utils-06cf0d4a27c3f798.rlib --extern futures=/usr/src/foundry/target/release/deps/libfutures-e27c8ea30c09d0a8.rlib --extern glob=/usr/src/foundry/target/release/deps/libglob-96e55c85069b1667.rlib --extern hex=/usr/src/foundry/target/release/deps/libhex-4b6a955ed4498af7.rlib --extern locate_cargo_manifest=/usr/src/foundry/target/release/deps/liblocate_cargo_manifest-04cf1e9c207dc0ee.rlib --extern once_cell=/usr/src/foundry/target/release/deps/libonce_cell-3a6d324dcc1a86f4.rlib --extern proptest=/usr/src/foundry/target/release/deps/libproptest-1d116b53bcabf7d8.rlib --extern rayon=/usr/src/foundry/target/release/deps/librayon-0c60f864b465c6b6.rlib --extern regex=/usr/src/foundry/target/release/deps/libregex-d1ea232f82b4b232.rlib --extern rpassword=/usr/src/foundry/target/release/deps/librpassword-afde6035b3715301.rlib --extern rustc_hex=/usr/src/foundry/target/release/deps/librustc_hex-7ff161deb5a27052.rlib --extern semver=/usr/src/foundry/target/release/deps
[2022-03-13T20:48:08.004Z] /libsemver-e70879db2fea9bbb.rlib --extern serde=/usr/src/foundry/target/release/deps/libserde-a952f9b94115863d.rlib --extern serde_json=/usr/src/foundry/target/release/deps/libserde_json-9b1582c76b4952aa.rlib --extern similar=/usr/src/foundry/target/release/deps/libsimilar-484d36495a81f210.rlib --extern solang_parser=/usr/src/foundry/target/release/deps/libsolang_parser-5ac32ff7dea64c78.rlib --extern tokio=/usr/src/foundry/target/release/deps/libtokio-36a1ff7474d60db5.rlib --extern tracing=/usr/src/foundry/target/release/deps/libtracing-33de1c3d6c0916e4.rlib --extern tracing_subscriber=/usr/src/foundry/target/release/deps/libtracing_subscriber-76606f35dd11a297.rlib --extern ui=/usr/src/foundry/target/release/deps/libui-581c8abb28dc7416.rlib --extern walkdir=/usr/src/foundry/target/release/deps/libwalkdir-647610b184c56b15.rlib --extern watchexec=/usr/src/foundry/target/release/deps/libwatchexec-e110c6394e61aa9b.rlib -L native=/usr/src/foundry/target/release/build/ring-4d04cc5a03b93a0a/out -L native=/usr/src/fo
[2022-03-13T20:48:08.005Z] undry/target/release/build/hidapi-rusb-52fe827113bd459d/out -L native=/usr/src/foundry/target/release/build/libusb1-sys-cc8491d2452ad21c/out -L native=/usr/src/foundry/target/release/build/libgit2-sys-5d083e97c2b8beff/out/build` (signal: 9, SIGKILL: kill)
#7 723.2 warning: build failed, waiting for other jobs to finish...
#7 1128.2 warning: `foundry-cli` (bin "forge") generated 1 warning
#7 1134.6 error: failed to compile `foundry-cli v0.1.0 (/usr/src/foundry/cli)`, intermediate artifacts can be found at `/usr/src/foundry/target`
#7 1134.6 
#7 1134.6 Caused by:
#7 1134.6   build failed
------
executor failed running [/bin/sh -c cd cli && cargo install --path .]: exit code: 101
[2022-03-13T20:48:08.179Z] Stop (1138614 ms): Run: docker build -f /Users/abdulrabbani/GitHub/abdulrabbani00/foundry/Dockerfile -t vsc-foundry-c9f4be51c7398a65a27f37432ee8d0a4 /Users/abdulrabbani/GitHub/abdulrabbani00/foundry
[2022-03-13T20:48:08.462Z] Command failed: docker build -f /Users/abdulrabbani/GitHub/abdulrabbani00/foundry/Dockerfile -t vsc-foundry-c9f4be51c7398a65a27f37432ee8d0a4 /Users/abdulrabbani/GitHub/abdulrabbani00/foundry
[2022-03-13T20:48:08.462Z] Exit code 1
```

I am branching off of `master`. Is there another branch I should be using? @mattsse

mattsse commented:
this is odd, 
> #7 723.1 error: could not compile `foundry-cli`; 1 warning emitted

not sure if this fails because of the warning

this is fixed on master rn

mind trying again?

abdulrabbani00 commented:
@mattsse - Already ahead of you, pulled the latest changes from master and re-running now. I will let you know how it goes but build is taking forever.

tynes commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM debian
+
+LABEL key="georgios@paradigm.xyz"
```
Is adding this label useful? Its a non standard label name for a maintainer, perhaps the label name should be maintainer like what the alpine image does, see https://hub.docker.com/r/alpine/git/dockerfile/

tynes commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+
+LABEL key="georgios@paradigm.xyz"
+
+RUN apt-get update -y; apt-get install -y curl git
```
You should add `&& rm -rf /var/lib/apt/lists/*` to this run line

tynes commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+RUN ["/bin/bash", "foundryup/install"]
+
+ENV PATH "$PATH:/root/.foundry/bin/"
+RUN echo "export PATH=/new/path:${PATH}" >> $HOME/.bashrc; \
```
Why is `/new/path` being added to the `$PATH`?

abdulrabbani00 replied:
The PATH needs to be added manually, it is not done on the install step.

tynes commented on Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM debian
```
Unsure if foundry can build on alpine, but `debian` is 124MB while `alpine` is 5.58MB. Definitely prefer building based on alpine if possible

abdulrabbani00 replied:
I've attempted to install it on alpine, but the build keeps failing due to dependency issues. Please feel free to try it on your own.

abdulrabbani00 commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,32 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
```
It's not necessary and I will remove it.

abdulrabbani00 commented on .github/workflows/release-container.yml:
```diff
@@ -0,0 +1,32 @@
+    runs-on: ubuntu-latest
+    steps:
+      -
+        name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
```
It is recommended as per the GHA documentation.

mattsse commented:
This was apparently introduced in the watchexec crate 

https://github.com/watchexec/watchexec/commit/465895ddecf7b9706dcfd28a5c84d65281e0e338

however our lockfile still pins notify to "5.0.0-pre.13"
So I wonder why this happens.
I'm not familiar with the inner workings of foundryup/install but I wonder why lock file is apparently ignored...
there's no new watchexec release on crates, we could patch until then but the lock file should prevent cases like this

mattsse commented:
Ah yes it's the  --locked param in cargo install at foundryup/install:131 which

     `--locked                    Require Cargo.lock is up to date`

can you try with removing this argument?

abdulrabbani00 commented:
Installing without `--locked` presented many more issues. 

I will work to ditch Debian and use `alpine` instead for our orgs needs. But I will also continue to use the `foundryup` command. I will attempt to compile locally but I am not sure that this will prove to be fruitful on `alpine` since it was not working on `debian`.

abdulrabbani00 commented:
@mattsse - Are you familiar with the following:

```
error[E0277]: expected a `FnMut<(Result<notify::Event, notify::Error>,)>` closure, found `Arc<std::sync::Mutex<impl notify::EventHandler>>`
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/watchexec-2.0.0-pre.11/src/fs.rs:51:57
   |
51 |             Self::Poll(delay) => notify::PollWatcher::with_delay(Arc::new(Mutex::new(f)), delay)
   |                                  ------------------------------- ^^^^^^^^^^^^^^^^^^^^^^^ expected an `FnMut<(Result<notify::Event, notify::Error>,)>` closure, found `Arc<std::sync::Mutex<impl notify::EventHandler>>`
   |                                  |
   |                                  required by a bound introduced by this call
   |
   = help: the trait `FnMut<(Result<notify::Event, notify::Error>,)>` is not implemented for `Arc<std::sync::Mutex<impl notify::EventHandler>>`
   = note: required because of the requirements on the impl of `EventHandler` for `Arc<std::sync::Mutex<impl notify::EventHandler>>`
note: required by a bound in `notify::PollWatcher::with_delay`
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/notify-5.0.0-pre.14/src/poll.rs:61:26
   |
61 |     pub fn with_delay<F: EventHandler>(event_handler: F, delay: Duration) -> Result<PollWatcher> {
   |                          ^^^^^^^^^^^^ required by this bound in `notify::PollWatcher::with_delay`

   Compiling foundry-cli v0.1.0 (/usr/src/foundry/cli)
For more information about this error, try `rustc --explain E0277`.
error: could not compile `watchexec` due to previous error
```

odyslam commented:
Why not foundryup @mattsse ?

abdulrabbani00 commented:
@gakonst , @mattsse - Is there a branch, tag, commit, or release that the team has high confidence in? Currently, when working off of the master branch the code does not compile. I believe I have resolved all system dependencies on `alpine`. But I still get various build errors as related to `watchexec` (see above).

abdulrabbani00 commented:
@gakonst and @mattsse - I have added a `Dockerfile` which resolves all the system dependencies for building the project from source on `alpine`. But the following error is still persisting.

```
   Compiling foundry-cli v0.1.0 (/usr/src/foundry/cli)
error[E0277]: expected a `FnMut<(Result<notify::Event, notify::Error>,)>` closure, found `Arc<std::sync::Mutex<impl notify::EventHandler>>`
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/watchexec-2.0.0-pre.11/src/fs.rs:51:57
   |
51 |             Self::Poll(delay) => notify::PollWatcher::with_delay(Arc::new(Mutex::new(f)), delay)
   |                                  ------------------------------- ^^^^^^^^^^^^^^^^^^^^^^^ expected an `FnMut<(Result<notify::Event, notify::Error>,)>` closure, found `Arc<std::sync::Mutex<impl notify::EventHandler>>`
   |                                  |
   |                                  required by a bound introduced by this call
   |
   = help: the trait `FnMut<(Result<notify::Event, notify::Error>,)>` is not implemented for `Arc<std::sync::Mutex<impl notify::EventHandler>>`
   = note: required because of the requirements on the impl of `EventHandler` for `Arc<std::sync::Mutex<impl notify::EventHandler>>`
note: required by a bound in `notify::PollWatcher::with_delay`
  --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/notify-5.0.0-pre.14/src/poll.rs:61:26
   |
61 |     pub fn with_delay<F: EventHandler>(event_handler: F, delay: Duration) -> Result<PollWatcher> {
   |                          ^^^^^^^^^^^^ required by this bound in `notify::PollWatcher::with_delay`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `watchexec` due to previous error
warning: build failed, waiting for other jobs to finish...
error: failed to compile `foundry-cli v0.1.0 (/usr/src/foundry/cli)`, intermediate artifacts can be found at `/usr/src/foundry/target`

Caused by:
  build failed
```

Any insight you can provide on the error would be helpful.

abdulrabbani00 commented:
@gakonst , @mattsse - I hope you guys are having a good Monday. Would you please consider merging the current PR, here are some of its strengths:

1. It contains a Dockerfile that will create an Alpine container and include `forge` and `cast`.
2. It contains another `Dockerfile` that can build the code from the source. This `Dockerfile` resolves all system dependencies. The only issue is that the current code base seems to have some compiler issues.
3. It contains a CI/CD pipeline that can be leveraged to cut new releases.

Although this PR is not perfect since it does not build the code from the source. I think it has many benefits:
1. It allows developers like me to have a working image of foundry to integrate into my codebase.
2. It contains a strong foundation for a future developer to leverage the `Dockerfile-from-source` to build the image from the source once the code is able to compile.

Cheers.

onbjerg commented on .github/workflows/release-container.yml:
```diff
@@ -0,0 +1,44 @@
+name: Create Docker Image
+
+on:
+  push:
+    tags:
+      - "v*.*.*"
+  schedule:
+    - cron: "0 0 * * *"
+  workflow_dispatch:
```
This will not always update to the latest nightly, since this action might be run before our publish step. Can we run this from the main release action using workflow calls? See https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_call

abdulrabbani00 replied:
I am not sure what your internal process/pipeline for release is, therefore it doesn't make much sense for me to poke around. Also I don't have more time to dedicate to this. I've gotten foundry installed in a docker container and this works for my team.

If you guys want to merge this code into some development branch and test it against your CI/CD pipeline, please feel free to do so. I think this can potentially make it easier for new users to quickly onboard to `foundry`.  At the very minimum, I would recommend maintaining an image that simply downloads the latest `forge` and `cast` binary.

Thank you

dmfxyz commented:
@abdulrabbani00 @gakonst @onbjerg @mattsse See a PR into @abdulrabbani00 's repo which builds foundry and forge from scratch using alpine. PRing into his repo as it builds off his work, and can probably re-use the github action. Final image size is 42.9MB

https://github.com/abdulrabbani00/foundry/pull/3/files

onbjerg commented:
Really appreciate the contribution @abdulrabbani00 and for kicking this off :smile: I understand you don't have more time to put into this PR, so I am closing it in favor of the work being done in #981 - but I also want to make it clear that I really do appreciate the contribution and conversation that has been had in here.

abdulrabbani00 commented:
@onbjerg - Happy to help in any way that I can. I would love the docker image in any way possible. As I know @dmfxyz has been doing the latest testing, perhaps he can chime in. I know he has done the last bit of work for this. 

Hope to see this merged soon :)
